llsd/serde_notation.py

```diff
@@ -210,97 +181,158 @@ def _parse_map(self):
+                    key = self._parse_string(cc)
                     found_key = True
                 elif cc.isspace() or cc == b',':
-                    self._getc()    # eat the character
+                    # ignore space or comma
                     pass
                 else:
                     self._error("Invalid map key")
             elif cc.isspace():
-                self._getc()    # eat the space
+                # ignore space
                 pass
             elif cc == b':':
-                self._getc()    # eat the ':'
-                value = self._parse()
+                # skip the ':'
+                value = self._parse(self._getc())
                 rv[key] = value
                 found_key = False
             else:
                 self._error("missing separator")
-            cc = self._peek()
-
-        if self._getc() != b'}':
-            self._error("Invalid map close token")
+            cc = self._getc()
 
         return rv
 
-    def _parse_array(self):
+    def _parse_array(self, cc):
         """
         parse a single array.
 
         array: [ object, object, object ]
         """
         rv = []
-        self._getc()    # eat the beginning '['
-        cc = self._peek()
+        # skip the beginning '['
+        cc = self._getc()
         while (cc != b']'):
             if cc is None:
                 self._error('Unclosed array')
             if cc.isspace() or cc == b',':
-                self._getc()
-                cc = self._peek()
+                cc = self._getc()
                 continue
-            rv.append(self._parse())
-            cc = self._peek()
+            rv.append(self._parse(cc))
+            cc = self._getc()
 
-        if self._getc() != b']':
-            self._error("Invalid array close token")
         return rv
 
-    def _parse_uuid(self):
+    def _parse_uuid(self, cc):
         "Parse a uuid."
-        self._getc()    # eat the beginning 'u'
+        # ignore the beginning 'u'
         # see comment on LLSDNotationFormatter.UUID() re use of latin-1
         return uuid.UUID(hex=self._getc(36).decode('latin-1'))
 
-    def _parse_uri(self):
+    def _parse_uri(self, cc):
         "Parse a URI."
-        self._getc()    # eat the beginning 'l'
-        return uri(self._parse_string())
+        # skip the beginning 'l'
+        return uri(self._parse_string(self._getc()))
 
-    def _parse_date(self):
+    def _parse_date(self, cc):
         "Parse a date."
-        self._getc()    # eat the beginning 'd'
-        datestr = self._parse_string()
+        # skip the beginning 'd'
+        datestr = self._parse_string(self._getc())
         return _parse_datestr(datestr)
 
-    def _parse_real(self):
+    def _parse_real(self, cc):
         "Parse a floating point number."
-        self._getc()    # eat the beginning 'r'
-        return float(self._get_re("real", _real_regex))
+        # recognize:
+        # [+-]?inf
+        # [+-]?nan
```

Comment:
Yes: https://github.com/secondlife/python-llsd/blob/main/llsd/serde_notation.py#L11 

I don't seriously expect to find `+nan` or `-nan` in existing serialized LLSD, but since the previous implementation accepted them and it would be impossible to prove no such instances exist...